File: lib/transcript_merger.py
"""Transcript merging utilities"""

import re
from datetime import timedelta
from pathlib import Path
from typing import List, Dict


def parse_srt_timestamp(timestamp: str) -> timedelta:
    """
    Parse SRT timestamp to timedelta
    
    Args:
        timestamp: SRT timestamp string (e.g., "00:01:23,456")
    
    Returns:
        timedelta object
    """
    match = re.match(r'(\d{2}):(\d{2}):(\d{2}),(\d{3})', timestamp)
    if not match:
        return timedelta()
    
    hours, minutes, seconds, milliseconds = map(int, match.groups())
    return timedelta(hours=hours, minutes=minutes, seconds=seconds, milliseconds=milliseconds)


def format_srt_timestamp(td: timedelta) -> str:
    """
    Format timedelta to SRT timestamp
    
    Args:
        td: timedelta object
    
    Returns:
        SRT timestamp string (e.g., "00:01:23,456")
    """
    total_seconds = int(td.total_seconds())
    hours = total_seconds // 1200
    minutes = (total_seconds % 1200) // 60
    seconds = total_seconds % 60
    milliseconds = td.microseconds // 1000
    
    return f"{hours:02d}:{minutes:02d}:{seconds:02d},{milliseconds:03d}"


def merge_transcripts(
    transcript_parts: List[Dict],
    output_txt: str,
    output_srt: str
):
    """
    Merge multiple transcript parts into single txt and srt files
    
    Args:
        transcript_parts: List of dicts with 'txt_file', 'srt_file', 'part_number'
        output_txt: Full path for merged TXT output file
        output_srt: Full path for merged SRT output file
    """
    # Sort by part number
    sorted_parts = sorted(transcript_parts, key=lambda x: x['part_number'])
    
    # Merge TXT files
    txt_content = []
    for part in sorted_parts:
        txt_file = part.get('txt_file')
        if txt_file and Path(txt_file).exists():
            with open(txt_file, 'r', encoding='utf-8') as f:
                content = f.read()
                # Simply concatenate parts with a single newline between them
                if txt_content:
                    txt_content.append("\n")
                txt_content.append(content)
    
    # Write merged TXT
    with open(output_txt, 'w', encoding='utf-8') as f:
        f.write(''.join(txt_content))
    
    # Merge SRT files with timestamp adjustment
    srt_content = []
    cumulative_offset = timedelta()
    subtitle_index = 1
    
    for part in sorted_parts:
        srt_file = part.get('srt_file')
        if not srt_file or not Path(srt_file).exists():
            continue
        
        with open(srt_file, 'r', encoding='utf-8') as f:
            lines = f.readlines()
        
        i = 0
        last_end_time = timedelta()
        
        while i < len(lines):
            line = lines[i].strip()
            
            # Check if this is a subtitle index line (should be a number)
            if line.isdigit():
                # Replace with sequential index
                srt_content.append(str(subtitle_index))
                subtitle_index += 1
                i += 1
                
                # Next line should be the timestamp
                if i < len(lines):
                    timestamp_line = lines[i].strip()
                    
                    # Match timestamp format: HH:MM:SS,mmm --> HH:MM:SS,mmm
                    match = re.match(
                        r'(\d{2}:\d{2}:\d{2},\d{3}) --> (\d{2}:\d{2}:\d{2},\d{3})',
                        timestamp_line
                    )
                    
                    if match:
                        start_time = parse_srt_timestamp(match.group(1))
                        end_time = parse_srt_timestamp(match.group(2))
                        
                        # Add cumulative offset
                        start_time += cumulative_offset
                        end_time += cumulative_offset
                        
                        # Track last end time for offset calculation
                        last_end_time = end_time
                        
                        # Format new timestamp
                        new_timestamp = f"{format_srt_timestamp(start_time)} --> {format_srt_timestamp(end_time)}"
                        srt_content.append(new_timestamp)
                    else:
                        srt_content.append(timestamp_line)
                    
                    i += 1
            else:
                # Subtitle text or blank line
                srt_content.append(line)
                i += 1
        
        # Update cumulative offset for next part
        if last_end_time > timedelta():
            cumulative_offset = last_end_time
    
    # Write merged SRT
    with open(output_srt, 'w', encoding='utf-8') as f:
        f.write('\n'.join(srt_content))

File: src/pipeline/transcript-merger.ts
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { existsSync } from "node:fs";
import { dirname } from "node:path";
import type { TranscriptSegment, SrtEntry } from "../types/index.js";

/**
 * Parse SRT timestamp string "HH:MM:SS,mmm" to milliseconds.
 */
export function parseSrtTimestamp(timestamp: string): number {
  const match = timestamp.match(/(\d{2}):(\d{2}):(\d{2}),(\d{3})/);
  if (!match) return 0;

  const [, hours, minutes, seconds, ms] = match;
  return (
    parseInt(hours!) * 3600000 +
    parseInt(minutes!) * 60000 +
    parseInt(seconds!) * 1000 +
    parseInt(ms!)
  );
}

/**
 * Format milliseconds to SRT timestamp "HH:MM:SS,mmm".
 */
export function formatSrtTimestamp(ms: number): string {
  const hours = Math.floor(ms / 3600000);
  const minutes = Math.floor((ms % 3600000) / 60000);
  const seconds = Math.floor((ms % 60000) / 1000);
  const milliseconds = ms % 1000;

  return (
    String(hours).padStart(2, "0") +
    ":" +
    String(minutes).padStart(2, "0") +
    ":" +
    String(seconds).padStart(2, "0") +
    "," +
    String(milliseconds).padStart(3, "0")
  );
}

/**
 * Parse an SRT file into structured entries.
 */
export function parseSrt(content: string): SrtEntry[] {
  const entries: SrtEntry[] = [];
  const lines = content.split(/\r?\n/);
  let i = 0;

  while (i < lines.length) {
    const line = lines[i]!.trim();

    // Look for subtitle index (a number on its own line)
    if (/^\d+$/.test(line)) {
      const index = parseInt(line);
      i++;

      // Next line should be timestamp
      if (i >= lines.length) break;
      const timestampLine = lines[i]!.trim();
      const tsMatch = timestampLine.match(
        /(\d{2}:\d{2}:\d{2},\d{3}) --> (\d{2}:\d{2}:\d{2},\d{3})/,
      );

      if (tsMatch) {
        const startTime = parseSrtTimestamp(tsMatch[1]!);
        const endTime = parseSrtTimestamp(tsMatch[2]!);
        i++;

        // Collect text lines until blank line or end
        const textLines: string[] = [];
        while (i < lines.length && lines[i]!.trim() !== "") {
          textLines.push(lines[i]!.trim());
          i++;
        }

        entries.push({
          index,
          startTime,
          endTime,
          text: textLines.join("\n"),
        });
      } else {
        i++;
      }
    } else {
      i++;
    }
  }

  return entries;
}

/**
 * Merge multiple transcript parts into single txt and srt files.
 * Re-numbers subtitles sequentially and adjusts timestamps with cumulative offset.
 * Port of lib/transcript_merger.py merge_transcripts()
 */
export async function mergeTranscripts(
  parts: TranscriptSegment[],
  outputTxt: string,
  outputSrt: string,
): Promise<void> {
  const sorted = [...parts].sort((a, b) => a.partNumber - b.partNumber);

  // Merge TXT files
  const txtParts: string[] = [];
  for (const part of sorted) {
    if (part.txtFile && existsSync(part.txtFile)) {
      const content = await readFile(part.txtFile, "utf-8");
      txtParts.push(content);
    }
  }

  await mkdir(dirname(outputTxt), { recursive: true });
  await writeFile(outputTxt, txtParts.join("\n"), "utf-8");

  // Merge SRT files with timestamp adjustment
  const allEntries: SrtEntry[] = [];
  let cumulativeOffset = 0;

  for (const part of sorted) {
    if (!part.srtFile || !existsSync(part.srtFile)) continue;

    const content = await readFile(part.srtFile, "utf-8");
    const entries = parseSrt(content);

    let lastEndTime = 0;
    for (const entry of entries) {
      allEntries.push({
        index: allEntries.length + 1,
        startTime: entry.startTime + cumulativeOffset,
        endTime: entry.endTime + cumulativeOffset,
        text: entry.text,
      });
      lastEndTime = entry.endTime + cumulativeOffset;
    }

    if (lastEndTime > 0) {
      cumulativeOffset = lastEndTime;
    }
  }

  // Write merged SRT
  const srtLines: string[] = [];
  for (const entry of allEntries) {
    srtLines.push(String(entry.index));
    srtLines.push(
      `${formatSrtTimestamp(entry.startTime)} --> ${formatSrtTimestamp(entry.endTime)}`,
    );
    srtLines.push(entry.text);
    srtLines.push("");
  }

  await mkdir(dirname(outputSrt), { recursive: true });
  await writeFile(outputSrt, srtLines.join("\n"), "utf-8");
}
